cloud/pkg/nodeupgradejobcontroller/controller/downstream.go

```diff
@@ -162,65 +178,121 @@ func (dc *DownstreamController) nodeUpgradeJobAdded(upgrade *v1alpha1.NodeUpgrad
+	if err != nil {
+		klog.Errorf("failed to drain node %s: %v", node, err)
+		return
+	}
+}
+
+// selectConcurrentNodes select the nodes to do upgrade operation, and put it into channel nodesChan
+func (dc *DownstreamController) selectConcurrentNodes(nodesChan chan string, allNodes []string, upgrade *v1alpha1.NodeUpgradeJob) {
+	// the default concurrency is 1
+	// this means that we will upgrade nodes one by one
+	// only when the last one node upgrade finished, we'll continue to upgrade the next one node
+	concurrency := 1
+	if upgrade.Spec.Concurrency != 0 {
+		concurrency = int(upgrade.Spec.Concurrency)
+	}
+
+	timeout := int(*upgrade.Spec.TimeoutSeconds) * (len(allNodes) + 1) / concurrency
+
+	err := wait.Poll(10*time.Second, time.Duration(timeout)*time.Second, func() (bool, error) {
+		upgradeJob, err := dc.crdClient.OperationsV1alpha1().NodeUpgradeJobs().Get(context.TODO(), upgrade.Name, metav1.GetOptions{})
 		if err != nil {
-			klog.Errorf("Failed to send upgrade message %v due to error %v", msg.GetID(), err)
-			continue
+			return false, nil
 		}
 
-		// process time out: cloud did not receive upgrade feedback from edge
-		// send upgrade timeout response message to upstream
-		go dc.handleNodeUpgradeJobTimeout(node, upgrade.Name, upgrade.Spec.Version, upgradeReq.HistoryID, upgrade.Spec.TimeoutSeconds)
-
-		// mark Upgrade state upgrading
-		status := &v1alpha1.UpgradeStatus{
-			NodeName: node,
-			State:    v1alpha1.Upgrading,
-			History: v1alpha1.History{
-				HistoryID:   upgradeReq.HistoryID,
-				UpgradeTime: time.Now().String(),
-			},
+		// if all the nodes upgrade is completed, close channel to inform that the upgrade is finished
+		if upgradeJob.Status.State == v1alpha1.Completed {
+			klog.Infof("all the nodes upgrade status are completed")
+			close(nodesChan)
+			return true, nil
 		}
-		err = patchNodeUpgradeJobStatus(dc.crdClient, upgrade, status)
-		if err != nil {
-			// not return, continue to mark node unschedulable
-			klog.Errorf("Failed to mark Upgrade upgrading status: %v", err)
+
+		// calculate the number of nodes in upgrading operation
+		upgradingNum := 0
+		for _, status := range upgradeJob.Status.Status {
+			if status.State == v1alpha1.Upgrading {
+				upgradingNum++
+			}
 		}
 
-		// mark edge node unschedulable
-		// the effect is like running cmd: kubectl drain <node-to-drain> --ignore-daemonsets
-		unscheduleNode := v1.Node{}
-		unscheduleNode.Spec.Unschedulable = true
-		// add a upgrade label
-		unscheduleNode.Labels = map[string]string{NodeUpgradeJobStatusKey: NodeUpgradeJobStatusValue}
-		byteNode, err := json.Marshal(unscheduleNode)
-		if err != nil {
-			klog.Warningf("marshal data failed: %v", err)
-			continue
+		// ensure the max number of upgrading nodes is Concurrency
+		if upgradingNum < concurrency {
+			for i := 0; i < concurrency-upgradingNum; i++ {
+				nodesChan <- allNodes[i]
+			}
+
+			allNodes = allNodes[:int(concurrency)-upgradingNum]
```

Comment:
allNodes = allNodes[int(concurrency)-upgradingNum:]？

Reply:
no, line 278, we put allNodes[0], allNodes[1],...,allNodes[int(concurrency)-upgradingNum-1] node to nodesChan, so we should remove the heading `int(concurrency)-upgradingNum` nodes before continue the nexe loop